File: backend/services/zfs_naming.py
"""Normalizzazione nomi pool/dataset/storage ZFS per replica syncoid."""

from __future__ import annotations

from typing import Optional


def collapse_dash_segments(name: str) -> str:
    """Rimuove segmenti consecutivi duplicati (es. replica-replica → replica)."""
    if not name:
        return name
    parts = name.split("-")
    out: list[str] = []
    for part in parts:
        if out and out[-1].lower() == part.lower():
            continue
        out.append(part)
    return "-".join(out)


def normalize_zfs_dest_path(
    pool: Optional[str],
    subfolder: Optional[str],
) -> tuple[str, Optional[str]]:
    """
    Separa pool ZFS e sottocartella evitando ridondanze nel nome pool.

    Es.: pool=ZFS-LARGE-replica + subfolder=replica → (ZFS-LARGE, replica)
    """
    clean_pool = (pool or "").strip().strip("/")
    clean_sub = (subfolder or "").strip().strip("/") or None
    if not clean_pool or not clean_sub:
        return clean_pool, clean_sub

    # Nome storage Proxmox già suffissato (ZFS-LARGE-replica)
    if "/" not in clean_pool and clean_pool.endswith(f"-{clean_sub}"):
        root = clean_pool[: -(len(clean_sub) + 1)]
        if root:
            return root, clean_sub

    # Path dataset già include la sottocartella (ZFS-LARGE/replica)
    if clean_pool.endswith(f"/{clean_sub}"):
        root = clean_pool[: -(len(clean_sub) + 1)].rstrip("/")
        if root:
            return root, clean_sub

    return clean_pool, clean_sub


def zfs_dataset_path(pool: str, subfolder: Optional[str]) -> str:
    pool, subfolder = normalize_zfs_dest_path(pool, subfolder)
    if not pool:
        return ""
    if subfolder:
        return f"{pool}/{subfolder}"
    return pool


def derive_zfs_storage_name(storage_name: Optional[str], zfs_pool: str) -> str:
    """Nome storage Proxmox coerente con il dataset ZFS, senza suffissi duplicati."""
    if not zfs_pool:
        return collapse_dash_segments((storage_name or "").strip())

    parts = zfs_pool.split("/")
    root = parts[0]
    base = (storage_name or root).strip()
    if len(parts) <= 1 or zfs_pool == base or zfs_pool == root:
        return collapse_dash_segments(base)

    sub = "-".join(parts[1:])
    from_root = f"{root}-{sub}"
    if base in (from_root, root):
        return collapse_dash_segments(from_root if base == root else base)
    if base.endswith(f"-{sub}"):
        return collapse_dash_segments(base)
    return collapse_dash_segments(f"{base}-{sub}")


def normalize_zfs_replica_dest(
    dest_pool: Optional[str],
    dest_subfolder: Optional[str],
    dest_storage: Optional[str] = None,
) -> tuple[str, Optional[str], str, str]:
    """
    Normalizza pool, subfolder, path dataset e nome storage destinazione.

    Returns: (pool, subfolder, zfs_path, storage_name)
    """
    pool, subfolder = normalize_zfs_dest_path(dest_pool, dest_subfolder)
    zfs_path = zfs_dataset_path(pool, subfolder)
    storage = derive_zfs_storage_name(dest_storage or pool, zfs_path)
    return pool, subfolder, zfs_path, storage

File: frontend/src/utils/zfsNaming.ts
/** Normalizzazione nomi pool/dataset/storage ZFS (allineato a backend/services/zfs_naming.py). */

export function collapseDashSegments(name: string): string {
  if (!name) return name
  const parts = name.split('-')
  const out: string[] = []
  for (const part of parts) {
    if (out.length && out[out.length - 1].toLowerCase() === part.toLowerCase()) continue
    out.push(part)
  }
  return out.join('-')
}

export function normalizeZfsDestPath(
  pool: string | null | undefined,
  subfolder: string | null | undefined,
): { pool: string; subfolder: string | null } {
  const cleanPool = (pool || '').trim().replace(/^\/+|\/+$/g, '')
  const cleanSub = (subfolder || '').trim().replace(/^\/+|\/+$/g, '') || null
  if (!cleanPool || !cleanSub) return { pool: cleanPool, subfolder: cleanSub }

  if (!cleanPool.includes('/') && cleanPool.endsWith(`-${cleanSub}`)) {
    const root = cleanPool.slice(0, -(cleanSub.length + 1))
    if (root) return { pool: root, subfolder: cleanSub }
  }
  if (cleanPool.endsWith(`/${cleanSub}`)) {
    const root = cleanPool.slice(0, -(cleanSub.length + 1)).replace(/\/+$/, '')
    if (root) return { pool: root, subfolder: cleanSub }
  }
  return { pool: cleanPool, subfolder: cleanSub }
}

export function zfsDatasetPath(pool: string, subfolder: string | null): string {
  const n = normalizeZfsDestPath(pool, subfolder)
  if (!n.pool) return ''
  return n.subfolder ? `${n.pool}/${n.subfolder}` : n.pool
}

export function deriveZfsStorageName(
  storageName: string | null | undefined,
  zfsPool: string,
): string {
  if (!zfsPool) return collapseDashSegments((storageName || '').trim())

  const parts = zfsPool.split('/')
  const root = parts[0]
  const base = (storageName || root).trim()
  if (parts.length <= 1 || zfsPool === base || zfsPool === root) {
    return collapseDashSegments(base)
  }

  const sub = parts.slice(1).join('-')
  const fromRoot = `${root}-${sub}`
  if (base === fromRoot || base === root) {
    return collapseDashSegments(base === root ? fromRoot : base)
  }
  if (base.endsWith(`-${sub}`)) return collapseDashSegments(base)
  return collapseDashSegments(`${base}-${sub}`)
}

export function normalizeZfsReplicaDest(
  destPool: string | null | undefined,
  destSubfolder: string | null | undefined,
  destStorage?: string | null,
) {
  const { pool, subfolder } = normalizeZfsDestPath(destPool, destSubfolder)
  const zfsPath = zfsDatasetPath(pool, subfolder)
  const storageName = deriveZfsStorageName(destStorage || pool, zfsPath)
  return { pool, subfolder, zfsPath, storageName }
}
